crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -991,7 +1001,54 @@ where
                 }
             }
             ast::Stmt::ImportFrom(node) => {
+                let mut found_star = false;
                 for (alias_index, alias) in node.names.iter().enumerate() {
+                    if &alias.name == "*" {
+                        // The following line maintains the invariant that every AST node that
+                        // implements `Into<DefinitionNodeKey>` must have an entry in the
+                        // `definitions_by_node` map. Maintaining this invariant ensures that
+                        // `SemanticIndex::definitions` can always look up the definitions for a
+                        // given AST node without panicking.
+                        //
+                        // The reason why maintaining this invariant requires special handling here
+                        // is that some `Alias` nodes may be associated with 0 definitions:
+                        // - If the import statement has multiple `*` names in the `names` list
+                        //   (e.g. `from foo import *, bar, *`)
+                        // - If the `*` import refers to a module that has 0 exported names.
+                        // - If the module being imported from cannot be resolved.
+                        self.add_entry_for_definition_key(alias.into());
+
+                        if found_star {
+                            continue;
+                        }
+
+                        found_star = true;
+
+                        // Wildcard imports are invalid syntax everywhere except the top-level scope,
+                        // and thus do not bind any definitions anywhere else
+                        if !self.current_scope_is_global_scope() {
+                            continue;
+                        }
+
+                        let Ok(module_name) =
+                            ModuleName::from_import_statement(self.db, self.file, node)
+                        else {
+                            continue;
+                        };
+
+                        let Some(module) = resolve_module(self.db, &module_name) else {
+                            continue;
+                        };
```

Comment:
I'm wondering if we should cache the resolved module in the `DefinitionKind` for `*` import definitions? It feels slightly silly to re-resolve the module name and module for these definitions in `infer.rs`, when we know that we wouldn't have created any definitions for the node during semantic indexing unless the module name and module were both resolvable.

Reply:
Hrm, this involves quite an annoying amount of refactoring unless we also cache the resolved `ModuleName` as well as the `File` on `StarImportDefinitionKind`. Leaning towards leaning this as-is for now, even though the duplication of logic between this file and `infer.rs` is somewhat annoying.